core/src/test/java/io/questdb/griffin/engine/functions/TouchTableFunctionTest.java

```diff
@@ -0,0 +1,152 @@
+    }
+
+    @Test
+    public void testTouchUpdateTouchAgain() throws Exception {
+        assertMemoryLeak(() -> {
+
+            final String query = "select touch(select * from x)";
+
+            final String ddl2 = "insert into x select * from (" +
+                    " select" +
+                    " rnd_geohash(40)," +
+                    " rnd_double(0)*100," +
+                    " 'VTJW'," +
+                    " to_timestamp('2019', 'yyyy') t" +
+                    " from long_sequence(100)" +
+                    ") timestamp (t)";
+
+            try {
+                execQuery(ddl, query);
+                execQuery(ddl2, query);
+            } catch (SqlException ex) {
+                Assert.fail(ex.getMessage());
+            }
+
+        });
+    }
+
+    @Test
+    public void testTouchTableTimeInterval() throws Exception {
+        assertMemoryLeak(() -> {
+            final String query = "select touch(select * from x where k in '1970-01-22')";
+            try {
+                execQuery(ddl, query);
+            } catch (SqlException ex) {
+                Assert.fail(ex.getMessage());
+            }
+        });
+    }
+
+    @Test
+    public void testTouchTableNoTimestampColumnSelected() throws Exception {
+        assertMemoryLeak(() -> {
+            final String query = "select touch(select g,a,b from x where k in '1970-01-22')";
+            try {
+                execQuery(ddl, query);
+            } catch (SqlException ex) {
+                TestUtils.assertContains(ex.getFlyweightMessage(), "query does not support framing execution and cannot be pre-touched");
+            }
+        });
+    }
+
+    @Test
+    public void testTouchTableThrowOnComplexFilter() throws Exception {
+        assertMemoryLeak(() -> {
+            final String query = "select touch(select * from x where k in '1970-01-22' and a > 100.0)";
+            try {
+                execQuery(ddl, query);
+            } catch (SqlException ex) {
+                TestUtils.assertContains(ex.getFlyweightMessage(), "query does not support framing execution and cannot be pre-touched");
+            }
+        });
+    }
+
+    @Test
+    public void testTouchTableTimeRange() throws Exception {
+        assertMemoryLeak(() -> {
+            final String query = "select touch(select * from x where k > '1970-01-18T00:00:00.000000Z')";
+            try {
+                execQuery(ddl, query);
+            } catch (SqlException ex) {
+                TestUtils.assertContains(ex.getFlyweightMessage(), "query does not support framing execution and cannot be pre-touched");
+            }
+        });
+    }
+
+    private void execQuery(String ddl, String query) throws SqlException {
+        compiler.compile(ddl, sqlExecutionContext);
+        CompiledQuery cc = compiler.compile(query, sqlExecutionContext);
+        printCursor(cc);
+    }
+
+    private void printCursor(CompiledQuery cc) throws SqlException {
```

Comment:
we have TestUtils.printSql() and TestUtils.assertSql()

Reply:
fixed